tiup/tiup-mirror.md

```diff
@@ -81,10 +92,117 @@ The `tiup mirror clone` command provides many optional flags (might provide more
+
+```bash
+export TIUP_MIRRORS=/shared_data/tiup
+tiup list
+```
+
+Using `TIUP_MIRRORS` can permanently change the mirror configuration, just like `tiup mirror set`. For details see [tiup issue #651](https://github.com/pingcap/tiup/issues/651).
+
+### Updating a private repo
+
+When running `tiup mirror clone` a second time with the same `target-dir` it will create new manifests and download newly available versions of components if applicable. It will skip downloading existing component versions.
+
+> **Note:**
+> 
+> As the manifest will be re-created make sure to include all components and all versions (including older ones that were previously downloaded) as otherwise they will be excluded from the manifest.
+
+## Custom repository
+
+A custom repository can be used with TiDB components like TiDB, TiKV or PD that you build yourself. It is also possible to create your own tiup components.
+
+Creating your own components can be done with the `tiup package` command. The instructions for how to do this can be found [here](https://github.com/pingcap/tiup/blob/master/doc/user/package.md).
+
+### Creating a custom repository
+
+To create an empty repository in `/data/mirror`:
+
+```bash
+tiup mirror init /data/mirror
+```
+
+As part of creating the repository keys will be written to `/data/mirror/keys`.
+
+To create a new private key in `~/.tiup/keys/private.json`:
+
+```bash
+tiup mirror genkey
+```
+
+Grant `jdoe` with private key `~/.tiup/keys/private.json` ownership of `/data/mirror`
+
+```bash
+tiup mirror set /data/mirror
+tiup mirror grant jdoe
+```
+
+### Working with custom components
+
+First we create a custom component called hello.
+
+```bash
+$ cat > hello.c << END
+> #include <stdio.h>
+int main() {
+  printf("hello\n");
+  return (0);
+}
+END
+$ gcc hello.c -o hello
+$ tiup package hello --entry hello --name hello --release v0.0.1
+```
+
+This creates `package/hello-v0.0.1-linux-amd64.tar.gz`.
+
+Then we create a new repository, a private key and then grant ownership to the repository.
+
+```bash
+$ tiup mirror init /tmp/m
+$ tiup mirror genkey
+$ tiup mirror set /tmp/m
+$ tiup mirror grant $USER
+```
+
+```bash
+tiup mirror publish hello v0.0.1 package/hello-v0.0.1-linux-amd64.tar.gz hello
+```
+
+And finally we run the component, if it wasn't installed yet it will be downloaded.
+
+```bash
+$ tiup hello
+```
+
+```
+The component `hello` version  is not installed; downloading from repository.
+Starting component `hello`: /home/dvaneeden/.tiup/components/hello/v0.0.1/hello
+hello
+```
+
+With `tiup mirror merge` you can merge a repository with custom components into another one. This assumes that all components in `/data/my_custom_components` are signed by the current `$USER`.
+
+```bash
+$ tiup mirror set /data/my_mirror
+$ tiup mirror grant $USER
+$ tiup mirror merge /data/my_custom_components
+```
```

Comment:
```suggestion
3. Run the component. If it is not installed yet, it will be downloaded first.

    ```bash
    $ tiup hello
    ```
    
    ```
    The component `hello` version  is not installed; downloading from repository.
    Starting component `hello`: /home/dvaneeden/.tiup/components/hello/v0.0.1/hello
    hello
    ```
    
    With `tiup mirror merge`, you can merge a repository with custom components into another one. This assumes that all components in `/data/my_custom_components` are signed by the current `$USER`.
    
    ```bash
    $ tiup mirror set /data/my_mirror
    $ tiup mirror grant $USER
    $ tiup mirror merge /data/my_custom_components
    ```
```